go-client/client/session_stub.go

```diff
@@ -0,0 +1,191 @@
+package client
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	sessionpb2 "github.com/deephaven/deephaven-core/go-client/internal/proto/session"
+	ticketpb2 "github.com/deephaven/deephaven-core/go-client/internal/proto/ticket"
+
+	"google.golang.org/grpc/metadata"
+)
+
+// tokenResp thread-safely protects the current session token (or an error in getting the session token).
+type tokenResp struct {
+	Lock  sync.Mutex
+	Token []byte
+	Error error
+}
+
+// getToken thread-safely returns the current token, or an error if an error has occurred at some point.
+func (tk *tokenResp) getToken() ([]byte, error) {
+	tk.Lock.Lock()
+	defer tk.Lock.Unlock()
+	if tk.Error != nil {
+		return nil, tk.Error
+	} else {
+		token := make([]byte, len(tk.Token))
+		copy(token, tk.Token)
+		return token, nil
+	}
+}
+
+// setToken thread-safely sets the session token to a new value.
+func (tk *tokenResp) setToken(tok []byte) {
+	tk.Lock.Lock()
+	tk.Token = tok
+	tk.Lock.Unlock()
+}
+
+// setError thread-safely sets an error value for the session token.
+func (tk *tokenResp) setError(err error) {
+	tk.Lock.Lock()
+	tk.Error = err
+	tk.Lock.Unlock()
+}
+
+// A refresher stores the current client token and sends periodic keepalive messages to refresh the client token.
+type refresher struct {
+	ctx         context.Context
+	sessionStub sessionpb2.SessionServiceClient
+
+	token *tokenResp // the actual client token, which gets periodically updated.
+
+	cancelCh <-chan struct{} // if this channel closes, the refresher should stop.
+
+	timeoutMillis int64 // time (in milliseconds) before the token should be refreshed again.
+}
+
+// refreshLoop is an endless loop that will update the token when necessary.
+// It can be canceled by closing the cancelCh channel.
+func (ref *refresher) refreshLoop() {
+	for {
+		select {
+		case <-ref.cancelCh:
+			return
+		case <-time.After(time.Duration(ref.timeoutMillis) * time.Millisecond):
+			err := ref.refresh()
+			if err != nil {
+				return
+			}
+		}
+	}
+}
+
+// startRefresher begins a background goroutine that will update the passed token whenever it would time out.
+func startRefresher(ctx context.Context, sessionStub sessionpb2.SessionServiceClient, token *tokenResp, cancelCh <-chan struct{}) error {
+	handshakeReq := &sessionpb2.HandshakeRequest{AuthProtocol: 1, Payload: [](byte)("hello godeephaven")}
```

Comment:
Does this payload have meaning?
If it does, is there a security risk from always sending the same first message?

Reply:
I do not know; I simply borrowed the message from the python API. @jmao-denver